### PR TITLE
Switch to using native support for CUDA instead of FindCUDA macros.

### PR DESCRIPTION
**Pull request by MarcusGDaniels**
### Summary

This switches to native support in CMake (versions >= 3.8) so that CUDA-specific CMake macros aren't used to set up the build.   It is to address issue #843 and make it possible to do per-CUDA file build options.

### CODE REVIEW

**Comment by MarcusGDaniels** [comment]
I reopened this pull request because the original one had a bad e-mail in it.  (One that I no longer have.)
Also a small change to this will address issue #787.  It amounts to including the current pybind11 headers in the source tree and adding a conditional in the bindings CMakefile to reference them.  It appears that Qiskit's license and pybind11's are compatible, perhaps at least for the purposes of a temporary workaround until there is a newer Conda install for pybind11.

File: CMakeLists.txt
Comment (vvilpas) [inline CMakeLists.txt:326]: It would be great for readability and maintainability to wrap in a function these kind of transformations that are used in several places and in different CMakeFiles

**Review by vvilpas — CHANGES_REQUESTED** [review]
Thanks @MarcusGDaniels to take care of this! I have started reviewing, leave some comments.

File: CMakeLists.txt
Comment (MarcusGDaniels) [reply CMakeLists.txt]: > I think it is safe to remove this now and leave the original code, as some default windows flags are now set above (line 131)

I checked this on a Windows build and it is still necessary for CUDA builds.  Presumably because it is not C++ as far as CMake is concerned.  (The addition of /bigobj is for C++ files only.)   I have a follow-up branch that tweaks the Pybind11 header file so that WIndows builds work, and I reinstalled that change there for now.

File: CMakeLists.txt
Comment (vvilpas) [inline CMakeLists.txt:247]: Have you ever experience a build failure saying something like "redefinition of std"? I get that error on one linux machine (but it runs fine on others).  It seems the flag `-std=c++14` is added twice on the `CUDA` compile commands: `nvcc .... -std=c++14 ... -std=c++14 ...`. So I removed the c++14 flag here and it compiles everything fine but fails at linkage with undefined symbols errors.

Comment (MarcusGDaniels) [reply CMakeLists.txt:247]: Haven't seen that.   I'll try removing it to see if it breaks my builds.

Comment (MarcusGDaniels) [reply CMakeLists.txt:247]: Builds work for me without -std-c++=14 in that spot on Ubuntu 20.04 LTS with CUDA 10.1.  There, I have to use -DCMAKE_CXX_COMPILER=g++-8.  Also it builds ok (again that way) on Windows 10 with CUDA 11. A guess is that the unresolved symbols arose by building half with one option, and another half with another option?

Comment (vvilpas) [reply CMakeLists.txt:247]: I don't really know. Don't worry, I'll check what is different on this other linux machine and try to make it work.

**Comment by vvilpas** [comment]
Hi @MarcusGDaniels. Can you merge this PR https://github.com/MarcusGDaniels/qiskit-aer/pull/2? With this commit we will temporarily enable our CI jobs to build GPU wheels. That way we can check that we will be able to build the pypi GPU package.

**Comment by MarcusGDaniels** [comment]
> Hi @MarcusGDaniels. Can you merge this PR [MarcusGDaniels#2](https://github.com/MarcusGDaniels/qiskit-aer/pull/2)? With this commit we will temporarily enable our CI jobs to build GPU wheels. That way we can check that we will be able to build the pypi GPU package.

Sure, just let me know when.

**Comment by vvilpas** [comment]
Good! Let's see if we are able to build the gpu packages.
Be careful when merging master again with this branch, because if we bring this commit with it, https://github.com/Qiskit/qiskit-aer/pull/998, we may accidentally publish the sdist wheel. 
To be safe, just revert my PR before merging master here.

**Comment by MarcusGDaniels** [comment]
> Good! Let's see if we are able to build the gpu packages.
> Be careful when merging master again with this branch, because if we bring this commit with it, #998, we may accidentally publish the sdist wheel.
> To be safe, just revert my PR before merging master here.

Looks like it can't find the CUDA compiler in that environment.

**Comment by MarcusGDaniels** [comment]
> Good! Let's see if we are able to build the gpu packages.
> Be careful when merging master again with this branch, because if we bring this commit with it, #998, we may accidentally publish the sdist wheel.
> To be safe, just revert my PR before merging master here.

Here's what I find:

1) it is necessary to get /usr/local/cuda/bin into the PATH somehow.  I hard-coded it for now.   I'm not sure where to insert this PATH change into the workflow manager (as opposed to CMakefiles).  

2) There this problem:

     + sh -c 'python3 /project/tools/verify_wheels.py'
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  for_each: failed to synchronize: cudaErrorNoDevice: no CUDA-capable device is detected

**Comment by MarcusGDaniels** [comment]
>     1. it is necessary to get /usr/local/cuda/bin into the PATH somehow.  I hard-coded it for now.   I'm not sure where to insert this PATH change into the workflow manager (as opposed to CMakefiles).

Ok, sorted out this problem.   Not sure what to do about runtime testing, though.  IIRC there was no solution GPU continuous integration?

**Comment by MarcusGDaniels** [comment]
The test looks ok except for the #2 issue of verify_wheels.py failing due to CUDA-capable devices being unavaiable.

**Comment by vvilpas** [comment]
hmm, not sure what is happening. @mtreinish can you help us with this? verify_wheel is failing with `CUDA-capable devices being unavailable` and @MarcusGDaniels had to add `nvcc` to the path.

**Comment by mtreinish** [comment]
> hmm, not sure what is happening. @mtreinish can you help us with this? verify_wheel is failing with `CUDA-capable devices being unavailable` and @MarcusGDaniels had to add `nvcc` to the path.

The GPU wheel job works fine today as is. You can see the build logs for the recent 0.7.0 release gpu wheels here: https://github.com/Qiskit/qiskit-aer/runs/1261468922?check_suite_focus=true#step:5:23 (although since they're so long you'll have to download them since the gha log viewer won't show the full log)

The `CUDA-capable devices being unavailable` error is weird because that means something is now trying to use a GPU, which shouldn't happen unless you explicitly set a GPU enabled simulator method like `statevector_gpu`. verify wheels doesn't specify a backend method: https://github.com/Qiskit/qiskit-aer/blob/master/tools/verify_wheels.py#L446 but I don't think automatic method selection will ever pick a gpu simulator.

As for the nvcc path that is also weird because I thought the nvidia cuda RHEL/CentOS packages we install as the first part in the job add it to the default environment path for us. Like the current cmake files find nvcc find and build using it without issue, so I'm not sure what would change there.

**Comment by MarcusGDaniels** [comment]
> As for the nvcc path that is also weird because I thought the nvidia cuda RHEL/CentOS packages we install as the first part in the job add it to the default environment path for us. Like the current cmake files find nvcc find and build using it without issue, so I'm not sure what would change there.

A change in this patch is getting away from the FundCUDA macros which guess at possible locations.

**Comment by MarcusGDaniels** [comment]
I'm doing a test run now that just disables verify_wheels since it would be redundant with the non-GPU build, and it can't in principle run anyway, right?  Unless, as you say it exercises the CPU code path.

**Comment by vvilpas** [comment]
I've been investigating a bit. At the beginning of execution the simulator checks which methods it can run. To do this it tries to simulate a small circuit for the different methods and checks if it was successful or not. If a method throws an exception, a catch block intercepts it and set the `success` field to `False`, setting the method to unavailable. This is what happened before when a 'gpu' method was selected and no device was available. Now, for some reason, we never reach the catch block and the program terminates when the exception is thrown.

**Comment by MarcusGDaniels** [comment]
> Now, for some reason, we never reach the catch block and the program terminates when the exception is thrown.

Is there a way to run the code as generated where the test would run and get a stacktrace?  (e.g. under gdb)

**Comment by vvilpas** [comment]
I've tried to get some info on the backtrace and in Release I haven't been able to clarify anything. 
So, I set up the docker image 'cibuildwheel' uses to do some debugging. Interestingly, in debug everything works as expected. I'll try to modify the code to get the same behavior as before.

**Comment by vvilpas** [comment]
I have a potential solution. I asked our CUDA experts and I am waiting for their feedback.

**Comment by vvilpas** [comment]
@MarcusGDaniels can you please merge https://github.com/MarcusGDaniels/qiskit-aer/pull/3? verify_tools should be working now with a small change I've made.

**Comment by MarcusGDaniels** [comment]
> @MarcusGDaniels can you please merge [MarcusGDaniels#3](https://github.com/MarcusGDaniels/qiskit-aer/pull/3)? verify_tools should be working now with a small change I've made.

I did this merge, but now I'm having trouble wit urllib3 being the incorrect version.   Not sure what happened.

**Comment by vvilpas** [comment]
> > @MarcusGDaniels can you please merge [MarcusGDaniels#3](https://github.com/MarcusGDaniels/qiskit-aer/pull/3)? verify_tools should be working now with a small change I've made.
> 
> I did this merge, but now I'm having trouble wit urllib3 being the incorrect version. Not sure what happened.

There is a compatibility problem between 2 conan dependencies: urllib3 and requests. They are fixing it and I've prepared a PR with a workaround in the meantime (#1031 ).

**Comment by vvilpas** [comment]
I have limited CUDA arch to just 5.2 to check that CI works. It seems we are running out of time with all the archs enabled.

**Comment by MarcusGDaniels** [comment]
> I have limited CUDA arch to just 5.2 to check that CI works. It seems we are running out of time with all the archs enabled.

Hmm, it's been running for hours.  I imagine it is stuck somehow but there is no log reported yet.

**Comment by MarcusGDaniels** [comment]
> > I have limited CUDA arch to just 5.2 to check that CI works. It seems we are running out of time with all the archs enabled.
> 
> Hmm, it's been running for hours. I imagine it is stuck somehow but there is no log reported yet.

Yeah, timed out on veryify_wheels.py again.

**Comment by vvilpas** [comment]
Yeah, I've already founded and hopefully fixed it (#1035). It was a change already introduced in master.

**Comment by MarcusGDaniels** [comment]
> Yeah, I've already found it and hopefully fixed it (#1035). It was a change already introduced in master.

Looked like it worked except for a Mac build that ran out of time.   Is there an automatic heuristic that retriggers the build or is there a way to force one without doing a commit?

**Comment by vvilpas** [comment]
> Looked like it worked except for a Mac build that ran out of time. Is there an automatic heuristic that retriggers the build or is there a way to force one without doing a commit?

Yeah, that was weird, but the full gpu build wheel action worked at last!!! 
I have disabled it so works again only for releases. If CI passes I think we can merge for upcoming release.

**Comment by MarcusGDaniels** [comment]
> > Looked like it worked except for a Mac build that ran out of time. Is there an automatic heuristic that retriggers the build or is there a way to force one without doing a commit?
> 
> Yeah, that was weird, but the full gpu build wheel action worked at last!!!
> I have disabled it so works again only for releases. If CI passes I think we can merge for upcoming release.

Reminder:  There's still a small patch to get Windows working.  It involves copying a modified pybind header file into the distribution, though.

**Comment by vvilpas** [comment]
> Reminder: There's still a small patch to get Windows working. It involves copying a modified pybind header file into the distribution, though.

No problem. I think we can address Win later as we don't officially support it yet. Pybind 2.6 is already released and fixes the error when compiling with nvcc.

**Review by chriseclectic — APPROVED** [review]

**Review by vvilpas — APPROVED** [review]
LGTM!

**Review by mtreinish — APPROVED** [review]

**Comment by mtreinish** [comment]
We probably should add a release note about this because it bumps the minimum cmake version. But we can do that after it's added to #1036

**Pull request merged**
